### ISSUE TITLE
Multiple requests made for refinement list

### ISSUE THREAD

**Issue by unclescooby** [typesense/typesense-instantsearch-adapter#10]
## Description

Faceted search not updating proper counts after clicking a refinement list.

## Steps to reproduce

Added html markup as per instructions and added refinementList widgets inside addWidgets(). If you can't reproduce then I will add full steps to reproduce.

## Expected Behavior

A single request with proper response/handling.

## Actual Behavior

After page load the initial request and result looks fine and html gets populated as expected. However, when clicking a value on a refinementList then 2 requests are sent. Clicking a value on another refinementList then 3 requests are sent. The request parameters differ and the results are all over the place.

## Metadata

**Typesense Version**: 0.15.0
**Typesense.js Version**: 0.7.0
**Typesense instantsearch adapter Version**: 0.3.0
**Instantsearch.js Version**: 4.5.0

**OS**: windows

**Comment by jasonbosco** [comment]
@unclescooby I haven't seen this issue with the datasets I've used. To help debug this further, could you create a minimal example (with sample data) on codesandbox.io replicating the issue?

**Comment by unclescooby** [comment]
I created a sandbox but never got round to setting up a secure Typesense server. In the meantime I've found the https://songs-search.typesense.org/ you recently set up, and it has the same behaviour.

Type something in the searchbox and the facets get enabled. Click on a facet value and 2 requests are sent, with different facet_by and filter_by parameters. First one shows all results and all facets, minus counts for "siblings" of selected values, which is what the second request returns. I do recall seeing inconsistencies with those values on my setup, but I'll recheck soon.

Can this not be done in a single request? An extra request for each facet gives undesired overhead.

**Comment by jasonbosco** [comment]
@unclescooby The request life cycle is fully managed by InstantSearch.js. What this adapter does is to transform the request parameters before sending it over to Typesense - it does not initiate its own requests. So the additional request that is being made comes from InstantSearch.js and is outside the control of the adapter. 

In any case, when you select a value in a refinement list, one of the requests fetches the actual search results - this is the one with `filter_by` set to the value that was chosen in the refinement list. Now, since we're filtering by a single value, the `facet_counts` field in the response from Typesense will only only contain counts for that value. But the refinementList widget also needs to update its UI with the counts for other values. So it makes a second request without any filters set and with `per_page=1` just to get facet counts for all values.

**Comment by unclescooby** [comment]
Ok, well I'm testing various search engines and I actually found typesense through a mention of one of the pages of meilisearch, so I think it would be ok to mention them. I've tested their implementation of instantsearch and it does not do extra requests for facets.

Anyway, thanks for explaining!  It makes me wonder if typesense can be modified to let filter_by filter the search results, but still return all applicable facet_counts, so that the extra requests aren't needed.

A workaround might be that if the results are first returned to the adapter, allowing for some preprocessing, then the first, unfiltered results are an opportunity to grab and store all facet values, then upon selecting a facet value the response could be supplemented with the missing counts, and instantsearch wouldn't need addional requests. Not sure if that could work and it adds complexity, but may be worthwhile to reduce load (imagine having 10 faceted fields!!)

**Comment by jasonbosco** [comment]
@unclescooby I finally got a chance to dig into why the adapter doesn't make the extra call when used with Algolia. It turns out, Algolia uses POST requests for search queries and actually batches the two queries (one for the facet values and another for the search results) into one POST request. Whereas with Typesense, search queries are done over GET and one-at-a-time without batching. So InstantSearch.js ends up automatically making two separate GET requests to Typesense. 

So the true fix for this issue would be to support batch search queries over POST in Typesense Server. I've created an issue for this in our backlog. 

> imagine having 10 faceted fields!!

Even if you have several faceted fields, the maximum number of requests that should be made is two - one for the search query and one to fetch all facets (even if there are multiple). Are you seeing network requests made for each facet separately?

**Comment by unclescooby** [comment]
Yes, I just visited the demo page again. Using google chrome on windows. In the dev tools network tab I'm filtering on xhr to prevent seeing xhr preflights and other requests.  Doing this: type e to reveal the facets (1 request), select a release date ie 1990s (+2 requests), check an artist (+3 requests), check a genre (+4 requests), check a release type (+5 requests), check a country (+6 requests).

I did not carefully inspect the request urls this time, just that the request is sent to xxxxxxx.typesense.net and I see the first request having all categories in the facet_by parameter, and the others only have one category in the facet_by parameter.

I also see the response times rising with each added selection... 3+ different facet categories it's getting slow....

**Comment by jasonbosco** [comment]
Ah you're right. I do see the multiple GET requests now - when refining by multiple fields, one request is made for each field, but with a different facet_by parameter. 

I also took a look at how Algolia handles it, it turns out they also make multiple requests - one for each field. So 10 filters actually make 10 _search_ requests in Algolia. *BUT*, the difference is that they allow for multiple search requests to be submitted together in one POST request. So though it shows up as one HTTP POST request, that one HTTP request actually has multiple Search requests within it.

So this confirms that the true solve for the multiple GET requests is for Typesense Server to also support multiple search requests in one POST request.

**Comment by jasonbosco** [comment]
@unclescooby We're adding support for multi_search in Typesense Server v0.19.0 (to be released soon). I updated the adapter to use that feature and now the adapter only makes 1 request for multiple search requests generated by InstantSearch! This update should be out in a week or two. 

Thank you for surfacing this!

**Comment by jasonbosco** [comment]
@Nick-Mazuk following up from https://github.com/typesense/typesense/issues/199#issuecomment-767802473 and https://github.com/typesense/typesense/issues/200: 

I have an RC build of v0.19.0 and a pre-release version of the typesense instantsearch adapter that uses the multi_search feature, which will reduce the number of HTTP requests that are made by the adapter. Would you be interested in beta testing it in your setup? 

@unclescooby Let me know if you'd be interested as well.

**Comment by Nick-Mazuk** [comment]
> @Nick-Mazuk following up from [typesense/typesense#199 (comment)](https://github.com/typesense/typesense/issues/199#issuecomment-767802473) and [typesense/typesense#200](https://github.com/typesense/typesense/issues/200):
> 
> I have an RC build of v0.19.0 and a pre-release version of the typesense instantsearch adapter that uses the multi_search feature, which will reduce the number of HTTP requests that are made by the adapter. Would you be interested in beta testing it in your setup?

Definitely! What would be the easiest way for me to help?

**Comment by jasonbosco** [comment]
Thanks @Nick-Mazuk! I just pushed out a pre-release version of the typesense instantsearch adapter as [`v0.4.0-0`](https://www.npmjs.com/package/typesense-instantsearch-adapter/v/0.4.0-0), which uses the multi-search feature in Typesense Server `v0.19.0`. Could you update your instant search to use this version?

Also, here's a one-line docker command to get the latest RC build of v0.19.0: 

```
docker run -i -p 8108:8108 -v/tmp/typesense-server-data/:/data typesense/typesense:0.19.0.rc9 --data-dir /data --api-key=xyz --listen-port 8108 --enable-cors
```

You should now see only one HTTP request for multiple refinementLists! Let me know if you run into issues.

**Comment by Nick-Mazuk** [comment]
> Thanks @Nick-Mazuk! I just pushed out a pre-release version of the typesense instantsearch adapter as [`v0.4.0-0`](https://www.npmjs.com/package/typesense-instantsearch-adapter/v/0.4.0-0), which uses the multi-search feature in Typesense Server `v0.19.0`. Could you update your instant search to use this version?
> 
> Also, here's a one-line docker command to get the latest RC build of v0.19.0:
> 
> ```
> docker run -i -p 8108:8108 -v/tmp/typesense-server-data/:/data typesense/typesense:0.19.0.rc9 --data-dir /data --api-key=xyz --listen-port 8108 --enable-cors
> ```
> 
> You should now see only one HTTP request for multiple refinementLists! Let me know if you run into issues.

So far, I've not had any issues with running v0.19 or the new adapter. I've actually converted several searches in my codebase to multi-search already!

I did have one issue installing the adapter, though. Essentially, the new `postinstall.js` is having issues running. I resorted to using `--force` on the install script to be able to download it. For reference, I got this error on the [demo repo](https://github.com/typesense/typesense-instantsearch-demo).

```bash
Error: Cannot find module '/path/to/repo/typesense-instantsearch-demo/node_modules/typesense-instantsearch-adapter/postinstall.js'
```

### Full install logs

```bash
$ npm i typesense-instantsearch-adapter@0.4.0-0 --force
npm WARN using --force I sure hope you know what you are doing.

> typesense-instantsearch-adapter@0.4.0-0 postinstall /path/to/repo/typesense-instantsearch-demo/node_modules/typesense-instantsearch-adapter
> node postinstall.js

internal/modules/cjs/loader.js:969
  throw err;
  ^

Error: Cannot find module '/path/to/repo/typesense-instantsearch-demo/node_modules/typesense-instantsearch-adapter/postinstall.js'
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:966:15)
    at Function.Module._load (internal/modules/cjs/loader.js:842:27)
    at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:71:12)
    at internal/main/run_main_module.js:17:47 {
  code: 'MODULE_NOT_FOUND',
  requireStack: []
}
npm WARN typesense@0.7.0 requires a peer of @babel/runtime@^7.10.2 but none is installed. You must install peer dependencies yourself.
npm WARN typesense-instantsearch-adapter@0.4.0-0 requires a peer of @babel/runtime@^7.10.2 but none is installed. You must install peer dependencies yourself.
npm WARN typesense@0.11.0-0 requires a peer of @babel/runtime@^7.10.2 but none is installed. You must install peer dependencies yourself.

+ typesense-instantsearch-adapter@0.4.0-0
added 5 packages from 4 contributors, removed 4 packages and audited 868 packages in 4.075s

31 packages are looking for funding
  run `npm fund` for details

found 2 high severity vulnerabilities
  run `npm audit fix` to fix them, or `npm audit` for details
```

And when not using force:

```bash
$ npm i typesense-instantsearch-adapter@0.4.0-0npm WARN rm not removing /path/to/repo/typesense-instantsearch-demo/node_modules/.bin/rimraf as it wasn't installed by /path/to/repo/typesense-instantsearch-demo/node_modules/rimrafnpm WARN rm not removing /path/to/repo/typesense-instantsearch-demo/node_modules/.bin/terser as it wasn't installed by/path/to/repo/typesense-instantsearch-demo/node_modules/terser
npm WARN rm not removing /path/to/repo/typesense-instantsearch-demo/node_modules/.bin/semver as it wasn't installed by /path/to/repo/typesense-instantsearch-demo/node_modules/semver
npm WARN rm not removing /path/to/repo/typesense-instantsearch-demo/node_modules/.bin/json5 as it wasn't installed by /path/to/repo/typesense-instantsearch-demo/node_modules/json5

> fsevents@1.2.13 install /path/to/repo/typesense-instantsearch-demo/node_modules/fsevents
> node install.js

  SOLINK_MODULE(target) Release/.node
  CXX(target) Release/obj.target/fse/fsevents.o
  SOLINK_MODULE(target) Release/fse.node

> deasync@0.1.20 install /path/to/repo/typesense-instantsearch-demo/node_modules/deasync
> node ./build.js

`darwin-x64-node-12` exists; testing
Binary is fine; exiting

> core-js@2.6.11 postinstall /path/to/repo/typesense-instantsearch-demo/node_modules/core-js
> node -e "try{require('./postinstall')}catch(e){}"


> parcel-bundler@1.12.4 postinstall /path/to/repo/typesense-instantsearch-demo/node_modules/parcel-bundler
> node -e "console.log('\u001b[35m\u001b[1mLove Parcel? You can now donate to our open collective:\u001b[22m\u001b[39m\n > \u001b[34mhttps://opencollective.com/parcel/donate\u001b[0m')"

Love Parcel? You can now donate to our open collective:
 > https://opencollective.com/parcel/donate

> typesense-instantsearch-adapter@0.4.0-0 postinstall /path/to/repo/typesense-instantsearch-demo/node_modules/typesense-instantsearch-adapter
> node postinstall.js

internal/modules/cjs/loader.js:969
  throw err;
  ^

Error: Cannot find module '/path/to/repo/typesense-instantsearch-demo/node_modules/typesense-instantsearch-adapter/postinstall.js'
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:966:15)
    at Function.Module._load (internal/modules/cjs/loader.js:842:27)
    at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:71:12)
    at internal/main/run_main_module.js:17:47 {
  code: 'MODULE_NOT_FOUND',
  requireStack: []
}
npm WARN typesense-instantsearch-adapter@0.4.0-0 requires a peer of @babel/runtime@^7.10.2 but none is installed. You must install peer dependencies yourself.
npm WARN typesense@0.11.0-0 requires a peer of @babel/runtime@^7.10.2 but none is installed. You must install peer dependencies yourself.

npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! typesense-instantsearch-adapter@0.4.0-0 postinstall: `node postinstall.js`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the typesense-instantsearch-adapter@0.4.0-0 postinstall script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/username/.npm/_logs/2021-01-27T03_13_50_964Z-debug.log
```

**Comment by jasonbosco** [comment]
Ah I forgot to add `postinstall.js` to the npm package. Just fixed it and published `v0.4.0-1`. Could you check now?

**Comment by unclescooby** [comment]
> @unclescooby Let me know if you'd be interested as well.
Thanks, I'm busy with other stuff right now, but will have a look soon. Good news though... this will reduce unneeded requests and roundtrips.

**Comment by Nick-Mazuk** [comment]
> Ah I forgot to add `postinstall.js` to the npm package. Just fixed it and published `v0.4.0-1`. Could you check now?

That fixed it! No other issues found yet and it's working great!

**Comment by Nick-Mazuk** [comment]
Hey @jasonbosco, I've run into a couple of segmentation faults today when using the `facet_query` parameter when using multi_search. Everything's been running smoothly until using that parameter, so that's my current best guess as to the cause. This is occurring with RC v0.19 of Typesense without using the instant-search adapter. Unfortunately, I have not been able to isolate which search query(s) is causing the issue.

Here's the stack trace.

```bash
E20210127 23:04:52.070109     1 backward.hpp:4192] Stack trace (most recent call last):
#17   Object "/opt/typesense-server", at 0x83abd4, in _start
#16   Object "/lib/x86_64-linux-gnu/libc-2.23.so", at 0x7f49fea3282f, in __libc_start_main
#15   Source "/typesense/src/main/typesense_server.cpp", line 124, in main [0x782840]
#14   Source "/typesense/src/typesense_server_utils.cpp", line 407, in run_server [0x937c29]
#13   Source "/typesense/src/http_server.cpp", line 193, in run [0x8e3470]
#12   Source "/typesense/external-Linux/h2o-5accc794404996fb056f1826f5a172693de95f22/lib/common/socket/evloop.c.h", line 619, in h2o_evloop_run [0x950bfe]
#11   Source "/typesense/external-Linux/h2o-5accc794404996fb056f1826f5a172693de95f22/lib/common/socket/evloop.c.h", line 572, in run_pending [0x94e907]
#10 | Source "/typesense/external-Linux/h2o-5accc794404996fb056f1826f5a172693de95f22/lib/common/socket/evloop.c.h", line 534, in read_on_ready
      Source "/typesense/external-Linux/h2o-5accc794404996fb056f1826f5a172693de95f22/lib/common/socket/evloop.c.h", line 248, in run_socket [0x94e7c8]
#90   Source "/typesense/external-Linux/h2o-5accc794404996fb056f1826f5a172693de95f22/lib/http1.c", line 638, in handle_incoming_request [0x965d6b]
#80   Source "/typesense/external-Linux/h2o-5accc794404996fb056f1826f5a172693de95f22/lib/http1.c", line 252, in h2o_socket_get_ptls [0x965f29]
#70   Source "/typesense/external-Linux/h2o-5accc794404996fb056f1826f5a172693de95f22/lib/http1.c", line 544, in write_req_non_streaming [0x966bf2]
#60   Source "/typesense/external-Linux/h2o-5accc794404996fb056f1826f5a172693de95f22/lib/core/request.c", line 154, in call_handlers [0x95d833]
#50   Source "/typesense/src/http_server.cpp", line 432, in catch_all_handler [0x8e70a7]
#40   Source "/typesense/src/http_server.cpp", line 536, in process_request [0x8e44d6]
#30   Source "/typesense/src/core_api.cpp", line 335, in post_multi_search [0x8d5762]
#20   Source "/typesense/src/collection_manager.cpp", line 751, in do_search [0x8bb752]
#10   Source "/typesense/src/collection.cpp", line 668, in search [0x87da89]
#00   Source "/typesense/include/string_utils.h", line 79, in trim [0x886647]
Segmentation fault (Signal sent by the kernel [(nil)])
```

**OS**: macOS 11.1

**Comment by jasonbosco** [comment]
@Nick-Mazuk Good catch! Here's an updated docker image that fixes this issue: 

```shell
docker run -i -p 8108:8108 -v/tmp/typesense-server-data/:/data typesense/typesense:0.19.0.rc11 --data-dir /data --api-key=xyz --listen-port 8108 --enable-cors
```

**Comment by Nick-Mazuk** [comment]
@jasonbosco , that seems to be working, no crashes yet this morning!

**Comment by Nick-Mazuk** [comment]
Hey @jasonbosco, with using the v0.19 beta over the last week, it seems fairly stable! However, I did notice another seg fault this morning. Opened up an issue on the main Typesense repo to keep this thread more on topic for the instant search adapter.

https://github.com/typesense/typesense/issues/212

**Comment by jasonbosco** [comment]
Just published `v0.19.0` of Typesense Server which supports multi_search. I've also published `v1.0.0` of the InstantSearch Typesense adapter which takes advantage of this feature. 

You should now see only one multi_search request being made regardless of the number of filter widgets used!
